src/classification/mondrian_tree.rs

```diff
@@ -0,0 +1,373 @@
+        if split_time > F::zero() {
+            // Here split the current node: if leaf we add children, otherwise
+            // we add a new node along the path
+            let feature = {
+                let cumsum = extensions
+                    .iter()
+                    .scan(F::zero(), |acc, &x| {
+                        *acc = *acc + x;
+                        Some(*acc)
+                    })
+                    .collect::<Array1<F>>();
+                let e_sample = F::from_f32(self.rng.gen::<f32>()).unwrap() * extensions.sum();
+                // DEBUG: shadowing with expected value
+                // let e_sample = F::from_f32(0.5).unwrap() * extensions.sum();
+                cumsum.iter().position(|&val| val > e_sample).unwrap()
+            };
+
+            let (lower_bound, upper_bound) = if x[feature] > node_min_list[feature] {
+                (
+                    node_min_list[feature].to_f32().unwrap(),
+                    x[feature].to_f32().unwrap(),
+                )
+            } else {
+                (
+                    x[feature].to_f32().unwrap(),
+                    node_max_list[feature].to_f32().unwrap(),
+                )
+            };
+            let threshold = F::from_f32(self.rng.gen_range(lower_bound..upper_bound)).unwrap();
+            // DEBUG: split in the middle
+            // let threshold = F::from_f32((lower_bound + upper_bound) / 2.0).unwrap();
+
+            let mut min_list = node_min_list.clone();
+            let mut max_list = node_max_list.clone();
+            min_list.zip_mut_with(x, |a, &b| *a = F::min(*a, b));
+            max_list.zip_mut_with(x, |a, &b| *a = F::max(*a, b));
+
+            // Create and push new parent node
+            let parent_node = Node {
+                parent: self.nodes[node_idx].parent,
+                time: self.nodes[node_idx].time,
+                is_leaf: false,
+                min_list,
+                max_list,
+                feature,
+                threshold,
+                left: None,
+                right: None,
+                stats: Stats::new(self.n_labels, self.n_features),
+            };
+
+            self.nodes.push(parent_node);
+            let parent_idx = self.nodes.len() - 1;
+            let sibling_idx = self.create_leaf(x, y, Some(parent_idx), split_time);
+
+            // Set the children appropriately
+            if x[feature] <= threshold {
+                // Grandpa: self.nodes[node_idx].parent
+                // (new) Parent: parent_idx
+                // Child: node_idx
+                // (new) Sibling: sibling_idx
+                self.nodes[parent_idx].left = Some(sibling_idx);
+                self.nodes[parent_idx].right = Some(node_idx);
+            } else {
+                self.nodes[parent_idx].left = Some(node_idx);
+                self.nodes[parent_idx].right = Some(sibling_idx);
+            }
```

Comment:
Sorry for my dumb questions 😅, I am learning Rust as I review your code.

I see that you created a parent node and might insert it above an existing decision node. And then there is a call to a `create_leaf` method to make a sibling node. Again, from what I get, you are creating an intermediate level above an existing node and assigning a sibling (which is a leaf) to this new node.

What if you are actually promoting a leaf to a decision node (an hence assigning new leaves as its descendants)?

Reply:
Very good point. I'll investigate it later today.

Reply:
This is the tree when we add a new parent (3) and a new sibling (4). Aren't we supposed to promote the sibling to a decision node? How would the tree look like otherwise?

```Rust
From:
┌─Node 1: left=Some(0), right=Some(2), parent=None, time=0.000, min=[0.2, 0.2], max=[0.4, 0.3], counts=[1, 1, 0]
│ ├─Node 0: left=None, right=None, parent=Some(1), time=1.228, min=[0.2, 0.2], max=[0.2, 0.2], counts=[1, 0, 0]
│ ├─Node 2: left=None, right=None, parent=Some(1), time=1.228, min=[0.4, 0.3], max=[0.4, 0.3], counts=[0, 1, 0]
To:
┌─Node 1: left=Some(0), right=Some(3), parent=None, time=0.000, min=[0.2, 0.2], max=[0.8, 0.8], counts=[1, 1, 0]
│ ├─Node 0: left=None, right=None, parent=Some(1), time=1.228, min=[0.2, 0.2], max=[0.2, 0.2], counts=[1, 0, 0]
│ ├─Node 3: left=Some(2), right=Some(4), parent=Some(1), time=1.228, min=[0.4, 0.3], max=[0.8, 0.8], counts=[0, 1, 1]
│ │ ├─Node 2: left=None, right=None, parent=Some(3), time=1.738, min=[0.4, 0.3], max=[0.4, 0.3], counts=[0, 1, 0]
│ │ ├─Node 4: left=None, right=None, parent=Some(3), time=1.738, min=[0.8, 0.8], max=[0.8, 0.8], counts=[0, 0, 1]
```

Reply:
The updated structure seems correct, at first glance.

However, I think I did not explain my original question correctly, so let me try to clarify it. 

I am having trouble finding the portion of the code that deals with the case where a leaf becomes a decision node. Unless I am missing something about Mondrian Trees, I believe they start with a single root node, and from there you can either:

1. Add a new parent to the current root (and, consequently, add a new sibling leaf). This seems to be the case in the first part of the example above
2. Promote the root (leaf) to a decision node and add some descendants (new leaves).

In the second case, the tree would look more or less like:

```
┌─Node 0: left=Some(1), right=Some(2), parent=None (...)
│ ├─Node 1: left=None, right=None, parent=Some(0) (...)
│ ├─Node 2: left=None, right=None, parent=Some(0) (...)
```

This reasoning should also be valid for subtrees rooted at an arbitrary node.

Am I getting something wrong?

Reply:
Yes you are right that is an issue. I'll be fixing it hopefully in the next few days after I fix a bigger problem related to enabling the sibling as decision node.